client/cli-derive/src/spec_factory.rs

```diff
@@ -0,0 +1,117 @@
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Configuration trait for a CLI based on substrate
+
+use proc_macro_error::{abort, abort_call_site};
+use quote::{quote, ToTokens};
+use std::collections::HashMap;
+use syn;
+use syn::parse::Parser;
+use syn::{punctuated::Punctuated, spanned::Spanned, *};
+
+pub(crate) fn spec_factory(
+	a: proc_macro::TokenStream,
+	i: proc_macro::TokenStream,
+) -> proc_macro2::TokenStream {
+	let attrs = match Punctuated::<ExprAssign, Token![,]>::parse_terminated.parse(a) {
+		Ok(x) => x,
+		Err(err) => abort!(err.span(), "could not parse attributes: {}", err),
+	};
+
+	let attrs = attrs
+		.iter()
+		.map(|x| {
+			(
+				match &*x.left {
+					Expr::Path(p) if p.path.segments.len() == 1 => {
+						p.path.segments[0].ident.to_string()
+					}
+					_ => abort!(x.left.span(), "could not parse token"),
+				},
+				(*x.right).to_token_stream(),
+			)
+		})
+		.collect::<HashMap<_, _>>();
+
+	let default_cli = quote! { Cli };
+	let cli = attrs.get("cli").unwrap_or(&default_cli);
+	let pkg_name = std::env::var("CARGO_PKG_NAME").unwrap().to_token_stream();
+	let impl_name = attrs.get("impl_name").unwrap_or(&pkg_name);
+	let default_impl_version = get_version().to_token_stream();
+	let impl_version = attrs.get("impl_version").unwrap_or(&default_impl_version);
+	let support_url = attrs
+		.get("support_url")
+		.unwrap_or_else(|| abort_call_site!("missing attribute: support_url"));
+	let executable_name = attrs.get("executable_name").unwrap_or(&pkg_name);
+	let default_author = std::env::var("CARGO_PKG_AUTHORS")
+		.unwrap_or_default()
+		.to_token_stream();
+	let author = attrs.get("author").unwrap_or(&default_author);
+	let default_description = std::env::var("CARGO_PKG_DESCRIPTION")
+		.unwrap_or_default()
+		.to_token_stream();
+	let description = attrs.get("description").unwrap_or(&default_description);
+	let copyright_start_year = attrs
+		.get("copyright_start_year")
+		.unwrap_or_else(|| abort_call_site!("missing attribute: copyright_start_year"));
+
+	let s: ItemFn = match syn::parse(i) {
+		Ok(x) => x,
+		_ => abort_call_site!("this macro only works on a function"),
+	};
+
+	quote! {
+		impl ::sc_cli::SubstrateCLI for #cli {
+			#s
+
+			fn get_impl_name() -> &'static str { #impl_name }
+			fn get_impl_version() -> &'static str { #impl_version }
+			fn get_support_url() -> &'static str { #support_url }
+			fn get_executable_name() -> &'static str { #executable_name }
+			fn get_author() -> &'static str { #author }
+			fn get_description() -> &'static str { #description }
+			fn get_copyright_start_year() -> i32 { #copyright_start_year }
+		}
+	}
+}
```

Comment:
Thanks! This whole macro is meant only to remove all the boilerplate with the VersionInfo. I renamed VersionInfo to SubstrateCLI because it now also has the load_spec/spec_factory function but if you have a better name I will just change it. Naming is hard

Reply:
I renamed the macro to substrate_cli_configuration